File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int totalProduct = 1;
        int zeroCount = 0;
        int[] result = new int[nums.length];
        for (int num : nums){
            if(num == 0){
                zeroCount++;
            }else{
                totalProduct = totalProduct * num;
            }
            
        }
        for (int i=0;i< nums.length;i++){
            if(zeroCount>0){
                if (nums[i] != 0){
                    result[i] = 0;
                }
                else{
                    if(zeroCount>1){
                        result[i] = 0;
                    }else{
                         result[i] = totalProduct;
                    }
                }
            }else{
                result[i] = totalProduct/nums[i];
            }
        }
        return result;
    }
}  

File: Data Structures & Algorithms/subsets-ii/submission-0.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        '''
        return all possible subsets without duplicates

        we also contains duplicate elements

        hence if the next element is also same, dont follow that route skip the element
        '''

        result = []
        state = []
        nums.sort()
        def dfs(i):
            nonlocal result,state
            # breaking condition
            if(i == len(nums)):
                result.append(state.copy())
                return
            if (i>=len(nums)):
                return
            

            #taking i
            
            state.append(nums[i])
            dfs(i+1)

            #without i
            state.pop()
            while(i+1<len(nums) and nums[i] == nums[i+1]):
                i = i+1
            dfs(i+1)
        
        dfs(0)
        return result
            #skip duplicate